File: main_project/DistCache/distributedcache/src/main/java/xce/distcache/PreferManager.java
package xce.distcache;

import java.util.List;
import java.util.Map;
import java.util.Random;

import xce.distcache.node.ReadNode;
import xce.distcache.node.Node.Item;

public class PreferManager {
    private Random _random;

    public PreferManager() {
	_random = new Random(System.currentTimeMillis());
    }

    public ReadNode choose(List<ReadNode> backends) {
	// TOOD complete this
	return backends.get(_random.nextInt(backends.size()));
    }

    public ReadNode choose(List<ReadNode> backends,
	    Map<ReadNode, List<Item>> categoried) {
	// TODO complete this using pref_data
	ReadNode current = backends.get(_random.nextInt(backends.size()));
	for (ReadNode b : backends) {
	    if (categoried.containsKey(b))
		return b;
	}
	return current;
    }
}

File: main_project/DistCache/distributedcache/src/main/cpp/PreferManager.cpp
#include "PreferManager.h"
#include "Node.h"

#include "LogWrapper.h"

using xce::distcache::node::ReadNodePtr;
using xce::distcache::node::ReadNode;
using xce::distcache::node::ItemPtr;
using xce::distcache::PreferManager;
using std::vector;
using std::map;

PreferManager::PreferManager()
{
	srand( time(NULL) );
}

ReadNodePtr PreferManager::choose(const vector<ReadNodePtr>& backends) {
	int bSize = (int)backends.size();

	if( bSize <= 0 )
	{
		bSize = 1;
		MCE_WARN( "PreferManger::choose WARN backends size = 0!" );
	}
	
	return backends.at( rand() % bSize );
}

ReadNodePtr PreferManager::choose(const vector<ReadNodePtr>& backends, const map<ReadNodePtr, vector<ItemPtr> >& categoried) {
	
	int iRandom = rand() % backends.size();
	ReadNodePtr current = ReadNodePtr( backends.at( iRandom ) );
	vector<ReadNodePtr>::const_iterator it = backends.begin();
	for ( vector<ReadNodePtr>::const_iterator it = backends.begin(); it != backends.end(); ++it )
	{
		if ( categoried.find( *it ) != categoried.end()) {
			return *it;
		}
	}
	return current;
}
